src/cubeb_pulse.c

```
@@ -170,46 +188,52 @@ stream_state_callback(pa_stream * s, void * u)
 }
 
 static void
-stream_request_callback(pa_stream * s, size_t nbytes, void * u)
+trigger_user_callback(pa_stream * s, void * input_data, size_t nbytes, cubeb_stream * stm)
 {
-  cubeb_stream * stm;
   void * buffer;
   size_t size;
   int r;
   long got;
-  size_t towrite;
+  size_t towrite, read_offset;
   size_t frame_size;
 
-  stm = u;
-
-  if (stm->shutdown)
-    return;
-
-  frame_size = WRAP(pa_frame_size)(&stm->sample_spec);
-
+  frame_size = WRAP(pa_frame_size)(&stm->output_sample_spec);
   assert(nbytes % frame_size == 0);
 
   towrite = nbytes;
-
+  read_offset = 0;
   while (towrite) {
     size = towrite;
     r = WRAP(pa_stream_begin_write)(s, &buffer, &size);
+    if (r < 0) {
+      // Never get here in normal scenario
+      LOG("Unexpected error. Debugging with rr causes this\n");
```

I am afraid the story is more complicated here. The bug experienced once by jesup while debugging with rr. The problem was that write stream was in state PA_STREAM_CREATING (so I guess the error was PA_ERR_BADSTATE) which should not be possible because we wait until stream is in PA_STREAM_READY state inside stream_init. Unfortunately I am not able to reproduce it here in order to track it and I prefer to check and report than failing silently.